### PR TITLE
Seperate console log to different endpoint

### PR DESCRIPTION
**Pull request by CheatCod**
# Description

Seperate console log to different endpoint

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

*Note: make sure your files are formatted with rust-analyzer*

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 ## vscode local history
 .history/
+.vscode
```
Comment (Ynng) [inline .gitignore]: we should probably keep .vscoed in the repo

File: src/handlers/events.rs [hunk 2]
```diff
@@ -29,22 +28,55 @@ pub async fn get_event_buffer(
...
+            .console_out_buffer
+            .lock()
+            .await
+            .get_ref()
+            .iter()
+            .rev()
+            .filter(|event| {
+                event.instance_uuid == uuid
+                    && matches!(event.event_inner, EventInner::InstanceOutput(_))
+                    && can_user_view_event(event, &requester)
```
Comment (Ynng) [inline src/handlers/events.rs:72]: your gonna filter it every time? just, a bit sketchy.

File: src/handlers/events.rs [hunk 3]
```diff
@@ -29,22 +28,55 @@ pub async fn get_event_buffer(
...
+            })
+            .cloned()
+            .collect(),
+    ))
 }
 
 #[derive(Deserialize)]
 pub struct WebsocketQuery {
     token: String,
+    uuid: String,
```
Comment (Ynng) [inline src/handlers/events.rs:82]: the overall "event_stream" also use this struct, but doesn't need the UUID
make them 2 structs instead?

File: src/handlers/events.rs [hunk 4]
```diff
@@ -94,3 +126,62 @@ async fn websocket(
...
+    let (mut sender, mut _receiver) = stream.split();
+    while let Ok(event) = event_receiver.recv().await {
+        if event.instance_uuid != uuid {
+            continue;
+        }
+        match event.event_inner {
+            EventInner::InstanceOutput(_) => {
+                if can_user_view_event(
+                    &event,
+                    match users.lock().await.get_ref().get(&uid) {
```
Comment (Ynng) [inline src/handlers/events.rs:178]: Can we just get the user once, instead of on every message?

**Review by Ynng — CHANGES_REQUESTED** [review]

File: src/handlers/events.rs (hunk 2 above)
Comment (CheatCod) [reply src/handlers/events.rs:72]: Wdym every time? it must be filtered

Comment (Ynng) [reply src/handlers/events.rs:72]: as in, instead of keeping the different instances' logs completely saparate

File: .gitignore (hunk 1 above)
Comment (CheatCod) [reply .gitignore]: Maybe in a different PR

File: src/handlers/events.rs (hunk 3 above)
Comment (CheatCod) [reply src/handlers/events.rs:82]: UUID is now required, and an uuid of "all" sends back all the events

File: src/handlers/events.rs (hunk 4 above)
Comment (CheatCod) [reply src/handlers/events.rs:178]: This is done for the following edge case:

1. User is authed and connected to WS
2. User is deleted before disconnecting
3. ??

In this case on the *next* event that will be send to the user, the WS connection will terminate

File: src/handlers/events.rs (hunk 3 above)
Comment (Ynng) [reply src/handlers/events.rs:82]: "all" feels sketchy but ok

File: .gitignore (hunk 1 above)
Comment (Ynng) [reply .gitignore]: well can you at least remove "adding .vscoed to gitignore" from this pr

File: src/handlers/events.rs (hunk 4 above)
Comment (Ynng) [reply src/handlers/events.rs:178]: sure... fine
although I still think maybe a dedicated "user_changed" event might be a better solution
AKA when something about a user changes, somewhere we emit a "user_changed" event

Comment (CheatCod) [reply src/handlers/events.rs:178]: Then each of the ws handler will need an event loop to notify them when a user changed, which is considerably more complex

Comment (Ynng) [reply src/handlers/events.rs:178]: fine

File: .gitignore (hunk 1 above)
Comment (CheatCod) [reply .gitignore]: removed in [2a27c8c](https://github.com/Lodestone-Team/client/pull/62/commits/2a27c8c0f3f852b03b9eaf795f873731834e7657)

**Review by Ynng — APPROVED** [review]

**Pull request merged**